### PR TITLE
[Sharktank][Llama][FP8] Minimal changes for numerically correct fp8

### PR DESCRIPTION
**Pull request by dan-garvey**
This patch enables the use of quark quantized models of the latest generation. Many changes were required to enable parity with the source model which is very sensitive to any numerical fluctuations. There is a test added to maintain this parity but is disabled until I can get the relevant data set up on the ci machine.

### CODE REVIEW

File: sharktank/sharktank/utils/patching.py [hunk 1]
```diff
@@ -55,11 +62,30 @@ def __init__(self):
...
+                orig_dup = self.tensors[name_base]
+                del self.tensors[name_base]
+                self.duplicate_tensors[name_base] = 0
+                self.tensors[f"{name_base}#0"] = orig_dup
+            elif name_base in self.duplicate_tensors:
+                index = self.duplicate_tensors[name_base] + 1
+                self.duplicate_tensors[name_base] = index
+                self.tensors[f"{name_base}#{index}"] = result_tensor
+            else:
+                self.tensors[name_base] = result_tensor
```
[the remark this answers is not in the document]
Comment (zjgarvey) [reply sharktank/sharktank/utils/patching.py:81]: Since it looks like this logic is used in `after_forward`, would it be useful to factor out something like the following?

```python
def insert_tensor(self, tensor : torch.Tensor, name_base : str):
   """ Adds a tensor to self.tensors while updating duplicate counts. """
```

[the remark this answers is not in the document]
Comment (zjgarvey) [reply sharktank/sharktank/utils/patching.py:81]: Actually, looking around more, I don't really see any instance where `name_base` would appear as one of the keys to `self.tensors`, so maybe the initial if statement is completely unnecessary?

File: sharktank/sharktank/utils/patching.py [hunk 2]
```diff
@@ -55,11 +62,30 @@ def __init__(self):
         # Map of module_name to last used index for duplicated tensors.
         self.duplicate_tensors = {}
 
+    def before_forward(self, module_name, module, *args, **kwargs):
+        for idx, arg in enumerate(args):
+            if not isinstance(arg, torch.Tensor):
+                continue
+            result_tensor = torch.detach(arg).contiguous().to(device="cpu").clone()
```
[the remark this answers is not in the document]
Comment (zjgarvey) [reply sharktank/sharktank/utils/patching.py:69]: Now that I think about it, perhaps adding this method warrants renaming the class to something like `SaveInputAndOutputTensorsPatch` and updating the docstrings for it.

**Review by zjgarvey — COMMENTED** [review]
I mostly have a bunch of naive questions, so I won't give approval or request changes.

File: sharktank/sharktank/examples/paged_llm_v1.py [hunk 3]
```diff
@@ -39,15 +39,14 @@ def __init__(
...
+        dump_bins: bool = False,
     ):
         self.model = model
         self.tokenizer = tokenizer
-        if self.model.config.kv_cache_type == "paged":
-            self.shared_cache_state = model.cache.allocate(page_cache_size)
-            self.free_pages = list(range(1, page_cache_size))
-        else:
-            self.shared_cache_state = None
+        self.shared_cache_state = model.cache.allocate(page_cache_size)
```
[the remark this answers is not in the document]
Comment (dan-garvey) [reply sharktank/sharktank/examples/paged_llm_v1.py:46]: We aren't supporting direct cache anymore

File: sharktank/sharktank/layers/linear.py [hunk 4]
```diff
@@ -88,7 +87,7 @@ def forward(self, x):
         # level to do this, but for now its here.
         if not isinstance(y, QuantizedTensor):
             if y.dtype == torch.float8_e4m3fnuz:
-                y = ops.to(y, torch.float16)
+                y = ops.to(y, torch.bfloat16)
```
[the remark this answers is not in the document]
Comment (dan-garvey) [reply sharktank/sharktank/layers/linear.py:90]: No this is an artifact of the way the model was quantized. The actual fp8 matmul intrinsic accumulates into f32, which iree can truncate, but in python we just cast to match the reference model

File: sharktank/sharktank/layers/paged_llama_attention_block.py [hunk 5]
```diff
@@ -128,22 +126,7 @@ def forward(
...
-                    .unpack()
-                    .dequant()
-                    .to(torch.float16)
-                )
-                xv = (
-                    self.cache_quantizer.quantize(xv)
-                    .unpack()
-                    .dequant()
-                    .to(torch.float16)
-                )
```
[the remark this answers is not in the document]
Comment (dan-garvey) [reply sharktank/sharktank/layers/paged_llama_attention_block.py:144]: Quark's model loader didn't support the fp8 kv cache. We are still doing it for export, but it is missing in the python comparison.

File: sharktank/sharktank/layers/paged_llama_attention_block.py [hunk 6]
```diff
@@ -175,11 +158,14 @@ def repeat_kv(x: torch.Tensor) -> torch.Tensor:
...
             xk = self.cache_quantizer.dequantize_raw_tensor(
-                xk, torch.float16, name="xk_deq"
+                xk, torch.bfloat16, name="xk_deq"
             )
             xv = self.cache_quantizer.dequantize_raw_tensor(
-                xv, torch.float16, name="xv_deq"
+                xv, torch.bfloat16, name="xv_deq"
             )
+            if attention_mask is not None:
+                attention_mask = attention_mask.to(torch.bfloat16)
```
[the remark this answers is not in the document]
Comment (dan-garvey) [reply sharktank/sharktank/layers/paged_llama_attention_block.py:167]: `if self.cache_quantizer and not self.fake_quant:`

File: sharktank/sharktank/models/llama/llama.py [hunk 7]
```diff
@@ -82,7 +86,7 @@ def __init__(self, theta: Theta, config: LlamaModelConfig):
 
         self.add_module(
             "token_embedding",
-            TokenEmbeddingLayer(theta("token_embd"), dtype=config.activation_dtype),
+            TokenEmbeddingLayer(theta("token_embd"), dtype=self.activation_dtype),
```
[the remark this answers is not in the document]
Comment (dan-garvey) [reply sharktank/sharktank/models/llama/llama.py:89]: Yeah this change is a no op. Style preference.

File: sharktank/sharktank/utils/export_artifacts.py [hunk 8]
```diff
@@ -94,6 +94,8 @@ def __init__(
         block_seq_stride: int,
         iree_hal_target_device: str,
         use_attention_mask: bool = False,
+        activation_dtype: str = "float16",
+        attention_dtype: str = "float16",
```
[the remark this answers is not in the document]
Comment (dan-garvey) [reply sharktank/sharktank/utils/export_artifacts.py:98]: most of our llama flock operates in this dtype. This is actually just maintaining that default for tests while adding the option to change it for the f8 model

File: sharktank/sharktank/utils/patching.py [hunk 9]
```diff
@@ -55,11 +62,30 @@ def __init__(self):
...
+            name_base = f"{module_name}_input_{idx}"
+            if name_base in self.tensors:
+                orig_dup = self.tensors[name_base]
+                del self.tensors[name_base]
+                self.duplicate_tensors[name_base] = 0
+                self.tensors[f"{name_base}#0"] = orig_dup
+            elif name_base in self.duplicate_tensors:
+                index = self.duplicate_tensors[name_base] + 1
+                self.duplicate_tensors[name_base] = index
```
[the remark this answers is not in the document]
Comment (dan-garvey) [reply sharktank/sharktank/utils/patching.py:78]: it does, but using index as a variable allows us to use it below in a way that is less ugly than a dictionary call

File: sharktank/tests/models/llama/quark_parity_test.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+
+from safetensors import safe_open
+import torch
+import unittest
+import pytest
+
+# is_mi300x = pytest.mark.skipif("config.getoption('iree_hip_target') != 'gfx942'")
```
Comment (aviator19941) [inline sharktank/tests/models/llama/quark_parity_test.py:13]: Remove commented code

File: sharktank/tests/models/llama/quark_parity_test.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+# @is_mi300x
```
Comment (aviator19941) [inline sharktank/tests/models/llama/quark_parity_test.py:15]: Remove commented code

**Review by aviator19941 — COMMENTED** [review]
Do you have a test or a command I can run to verify the numerics are correct?

File: sharktank/sharktank/utils/export_artifacts.py [hunk 12]
```diff
@@ -3,7 +3,7 @@
 # Licensed under the Apache License v2.0 with LLVM Exceptions.
 # See https://llvm.org/LICENSE.txt for license information.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-
+import torch
```
Comment (archana-ramalingam) [inline sharktank/sharktank/utils/export_artifacts.py:6]: Not used can be removed.

File: sharktank/sharktank/models/llama/llama.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (archana-ramalingam) [reply sharktank/sharktank/models/llama/llama.py:89]: For consistency, we can use `self` or `config`

File: sharktank/sharktank/layers/configs/llm_configs.py [hunk 13]
```diff
@@ -49,7 +49,7 @@ def from_gguf_props(p: dict[str, Any]):
         name_prefix = p.get("general.architecture", "llama")
         default_expert_count = 0
         default_expert_used_count = 0
-        default_rope_freq_base = 10000.0
+        default_rope_freq_base = 500000.0
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/configs/llm_configs.py:52]: Default value is 10000 for most models. Can we pass this separately?

File: sharktank/sharktank/examples/paged_llm_v1.py [hunk 14]
```diff
@@ -160,6 +157,23 @@ def prefill(self):
             attention_mask = replicate(attention_mask, tp)
             seq_block_ids_tensor = replicate(seq_block_ids_tensor, tp)
 
+        if self.dump_bins:
```
Comment (archana-ramalingam) [inline sharktank/sharktank/examples/paged_llm_v1.py:160]: You can use [generate_data.py](https://github.com/nod-ai/shark-ai/blob/1ef5b1fd466e9bd94bf9c2759820fb0e56fa576e/sharktank/sharktank/models/llama/tools/generate_data.py#L10) to fetch input data, given a model & a prompt.

File: sharktank/sharktank/layers/configs/llm_configs.py (hunk 13 above)
Comment (archana-ramalingam) [reply sharktank/sharktank/layers/configs/llm_configs.py:52]: Perplexity seems to be passing, so not a blocker.

File: sharktank/sharktank/ops/qlinear_impls.py [hunk 15]
```diff
@@ -52,9 +52,7 @@ def qlinear_tensor_scaled(
     if x_layout.qs.dtype.is_floating_point or weight_layout.qs.dtype.is_floating_point:
         if x_layout.qs.dtype == torch.float8_e4m3fnuz:
             # assume quark
-            return matmul(x_layout.qs, weight_layout.qs, transpose_rhs=True).to(
-                torch.float16
-            )
+            return matmul(x_layout.qs, weight_layout.qs, transpose_rhs=True)
```
Comment (nithinsubbiah) [inline sharktank/sharktank/ops/qlinear_impls.py:55]: If we do a return here, we're not actually inserting the mmt kernel which is what's intended from this script.

Comment (dan-garvey) [reply sharktank/sharktank/ops/qlinear_impls.py:55]: what does the mmt kernel do? because we can just lower fp8 matmul with torch

File: sharktank/sharktank/utils/export_artifacts.py (hunk 12 above)
Comment (dan-garvey) [reply sharktank/sharktank/utils/export_artifacts.py:6]: ah good catch thanks

**Comment by dan-garvey** [comment]
> Do you have a test or a command I can run to verify the numerics are correct?

you can run pytest sharktank/tests/models/llama/quark_parity_test.py on mi300x right now. I plan to make it run anywhere in a follow up patch

File: sharktank/sharktank/layers/configs/llm_configs.py (hunk 13 above)
Comment (dan-garvey) [reply sharktank/sharktank/layers/configs/llm_configs.py:52]: llama3 defaults to 500000 so I think we should use that

File: sharktank/sharktank/examples/paged_llm_v1.py (hunk 14 above)
Comment (IanNod) [reply sharktank/sharktank/examples/paged_llm_v1.py:160]: Can we update generate_data.py to use torch tensors instead of numpy arrays? Should work around this issue

File: sharktank/sharktank/layers/configs/llm_configs.py (hunk 13 above)
Comment (IanNod) [reply sharktank/sharktank/layers/configs/llm_configs.py:52]: Do we know where that 10000 came from?

File: sharktank/sharktank/layers/linear.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (IanNod) [reply sharktank/sharktank/layers/linear.py:90]: I'm not a fan that the python implementation can really only compare for one specific quantization method like this. I don't have an answer off the top of my head, so fine for now but ideally would be good to make more agnostic somehow

File: sharktank/tests/models/llama/benchmark_amdgpu_test.py [hunk 16]
```diff
@@ -279,7 +281,11 @@ def testBenchmark8B_f16_Non_Decomposed_Input_Len_2048(self):
             cwd=self.repo_root,
         )
 
-    @pytest.mark.xfail(reason="Compile Error", strict=True, raises=IreeCompileException)
+    @pytest.mark.xfail(
+        reason="Benchmark inputs not configured yet.",
```
Comment (IanNod) [inline sharktank/tests/models/llama/benchmark_amdgpu_test.py:285]: Can you add a todo and assign yourself? I would like to ensure this is done quickly to track and prevent regressions

**Review by IanNod — COMMENTED** [review]
Mostly looks good to me, some minor nit comments

**Comment by IanNod** [comment]
> > Do you have a test or a command I can run to verify the numerics are correct?
> 
> you can run pytest sharktank/tests/models/llama/quark_parity_test.py on mi300x right now. I plan to make it run anywhere in a follow up patch

Would be great if you could add to documentation so it's easy for anyone else to pick up and test/use fp8 models

**Comment by dan-garvey** [comment]
https://github.com/nod-ai/shark-ai/issues/880
https://github.com/nod-ai/shark-ai/issues/881
@IanNod

**Comment by dan-garvey** [comment]
I'll add documentation on halo models this afternoon

File: sharktank/sharktank/utils/patching.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (zjgarvey) [reply sharktank/sharktank/utils/patching.py]: Ah, yeah. Makes sense.

File: sharktank/sharktank/layers/paged_llama_attention_block.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zjgarvey) [reply sharktank/sharktank/layers/paged_llama_attention_block.py:167]: Yeah, that's what I'm asking about. Do we need `cache_quantizer` and `not fake_quant` for the attention mask to be in `bfloat16`? I guess that's probably the case.

**Comment by dan-garvey** [comment]
10000 is the default for llama2

File: sharktank/sharktank/layers/configs/llm_configs.py (hunk 13 above)
Comment (dan-garvey) [reply sharktank/sharktank/layers/configs/llm_configs.py:52]: llama2

**Review by IanNod — APPROVED** [review]
